lambda/src/server/main.cljs

```diff
@@ -28,23 +30,31 @@
   "Main AWS Lambda handler. Invoked by slackBot.
    See https://docs.aws.amazon.com/lambda/latest/dg/nodejs-handler.html"
   [^:js {:as req :keys [body]} ^js res]
-  (p/let [{:as result
-           :keys [response
-                  task]} (cond
-                           ;; Slack API: identification challenge
-                           (:challenge body) {:response (:challenge body)}
-
-                           ;; Slack Interaction (e.g. global shortcut)
-                           (:payload body) (handlers/handle-interaction! (:payload body))
-
-                           ;; Slack Event
-                           (:event body) (handlers/handle-event! (:event body)))]
-
-    (assert (or (map? result) (nil? result)))
-    (pp/pprint [(if result ::handled ::not-handled) body])
-    (when task (tasks/publish! task))
-
-    (.send res response)))
+  (if (:challenge body)
+    (.send res (:challenge body))
+    (p/let [[kind data team-id] (cond (:payload body) [:interaction (:payload body) (-> body :payload :team :id)]
+                                      (:event body) [:event (:event body) (:team_id body)]
+                                      (:challenge body) [:challenge (:challenge body) nil])
+            token (some-> team-id slack-db/team->token)
+            {:as result
+             :keys [response
+                    task]} (case kind
+                             ;; Slack API: identification challenge
+                             :challenge {:response data}
+
+                             ;; Slack Interaction (e.g. global shortcut)
+                             :interaction (handlers/handle-interaction! token data)
+
+                             ;; Slack Event
+                             :event (handlers/handle-event! token data))]
+
+      (assert (or (map? result) (nil? result)))
+      (pp/pprint [(if result ::handled ::not-handled) body])
+      (when task
+        (pp/pprint task)
+        (tasks/publish! task))
+
+      (.send res response))))
```

Comment:
So this is the immediate `200/OK`, and `tasks/publish!` handles actually doing something, correct?

Reply:
Right. the kitchen-async `p/let` has an implicit `p/do` for its body, meaning it will wait for any promises in the body to resolve before continuing, so the response is sent as soon as the task is published to the topic.